File: ade-basirwfrd/child-monitor-server/android/ChildMonitor/app/src/main/java/com/example/childmonitor/JudiFilter.java
package com.example.childmonitor;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * Client-side gambling/blocked site filter.
 * Checks against BOTH hardcoded keywords AND a dynamic blocklist synced from the server.
 */
public class JudiFilter {

    private static final String TAG = "JudiFilter";

    // Dynamic blocklist synced from server
    private static final Set<String> serverBlocklist = new HashSet<>();
    private static long lastSyncTime = 0;
    private static final long SYNC_INTERVAL = 60 * 1000L; // Sync every 60 seconds

    private static final List<String> JUDI_KEYWORDS = Arrays.asList(
            "judi", "slot", "poker", "togel", "casino", "bet", "bola",
            "gambling", "taruhan", "bandar", "jackpot", "scatter", "gacor",
            "maxwin", "pragmatic", "pgsoft", "livecasino", "sportsbook",
            "roulette", "blackjack", "baccarat", "sicbo", "judol",
            "slotgacor", "cuan",
            "sbobet", "mansion88", "m88", "w88", "fun88", "we88", "md88",
            "bk8", "dafabet", "cmd368", "ibcbet", "maxbet", "188bet",
            "longfu88", "kawanslot", "naga303", "dewapoker", "idnpoker",
            "joker123", "mpo88", "parimatch", "paripesa", "1xbet", "1win",
            "mostbet", "melbet", "roobet", "bcgame"
    );

    /** Fallback lokal (anti-judol saja — bukan antibokep). Full list dari server /api/blocklist. */
    private static final List<String> JUDI_DOMAINS = Arrays.asList(
            "sbobet.com", "ibcbet.com", "maxbet.com", "188bet.com",
            "bet365.com", "1xbet.com", "22bet.com", "melbet.com",
            "m88.com", "w88.com", "fun88.com", "we88.com", "md88.com",
            "dafabet.com", "cmd368.com", "bk8.com", "bk8.plus", "longfu88.com",
            "poker88.com", "idnpoker.com", "dewapoker.com", "pokerstars.com",
            "togel.com", "togel4d.com", "totobet.com", "indotogel.com",
            "slot88.com", "joker123.com", "joker388.com", "slotgacor.com",
            "judionline.com", "judibola.com", "pragmaticplay.com", "pgsoft.com",
            "bonanza88.com", "mpo88.com", "ole777.com", "stake.com", "bc.game",
            "mostbet.com", "kawanslot.com", "naga303.com", "parimatch.com",
            "1win.com", "roobet.com", "bheestybaulk.top", "aungudie.com", "zlink.fun"
    );

    /**
     * Sync blocklist from server. Call this periodically.
     */
    public static void syncBlocklist(Context context) {
        long now = System.currentTimeMillis();
        if (now - lastSyncTime < SYNC_INTERVAL) return; // Don't sync too often
        lastSyncTime = now;

        try {
            RetrofitClient.getInstance(context).getBlocklist().enqueue(new Callback<BlocklistResponse>() {
                @Override
                public void onResponse(Call<BlocklistResponse> call, Response<BlocklistResponse> response) {
                    if (response.isSuccessful() && response.body() != null && response.body().domains != null) {
                        synchronized (serverBlocklist) {
                            serverBlocklist.clear();
                            for (String d : response.body().domains) {
                                serverBlocklist.add(d.toLowerCase());
                            }
                        }
                        Log.d(TAG, "Blocklist synced: " + serverBlocklist.size() + " domains from server");
                    }
                }

                @Override
                public void onFailure(Call<BlocklistResponse> call, Throwable t) {
                    Log.w(TAG, "Blocklist sync failed: " + t.getMessage());
                }
            });
        } catch (Exception e) {
            Log.w(TAG, "Blocklist sync error: " + e.getMessage());
        }
    }

    /**
     * Check if a URL is a gambling/blocked site
     */
    public static boolean isJudiSite(String url) {
        if (url == null || url.isEmpty()) return false;

        String urlLower = url.toLowerCase();
        String host = extractHost(urlLower);

        // 1. Check against hardcoded domain list
        if (host != null) {
            for (String domain : JUDI_DOMAINS) {
                if (host.equals(domain) || host.endsWith("." + domain)) {
                    Log.d(TAG, "BLOCKED (hardcoded domain): " + url);
                    return true;
                }
            }
        }

        // 2. Check against server-synced blocklist
        if (host != null) {
            synchronized (serverBlocklist) {
                for (String domain : serverBlocklist) {
                    if (host.equals(domain) || host.endsWith("." + domain)) {
                        Log.d(TAG, "BLOCKED (server blocklist): " + url);
                        return true;
                    }
                }
            }
        }

        // 3. Check keywords in hostname
        if (host != null) {
            for (String keyword : JUDI_KEYWORDS) {
                if (host.contains(keyword)) {
                    Log.d(TAG, "BLOCKED (keyword in host): " + url);
                    return true;
                }
            }
        }

        // 4. Check specific keywords in URL path
        String[] pathKeywords = {"slot", "togel", "casino", "poker", "sportsbook"};
        for (String keyword : pathKeywords) {
            if (urlLower.contains("/" + keyword)) {
                Log.d(TAG, "BLOCKED (keyword in path): " + url);
                return true;
            }
        }

        return false;
    }

    private static String extractHost(String url) {
        try {
            String host = url;
            if (host.contains("://")) {
                host = host.substring(host.indexOf("://") + 3);
            }
            if (host.contains("/")) {
                host = host.substring(0, host.indexOf("/"));
            }
            if (host.contains(":")) {
                host = host.substring(0, host.indexOf(":"));
            }
            return host;
        } catch (Exception e) {
            return null;
        }
    }
}

File: ade-basirwfrd/child-monitor-server/backend/utils/judiFilter.js
const { getAllBlockedDomains, insertBlockedDomain, deleteBlockedDomain } = require('../models/db');

// Keywords yang mengindikasikan situs judi
const JUDI_KEYWORDS = [
    'judi', 'slot', 'poker', 'togel', 'casino', 'bet', 'bola',
    'gambling', 'taruhan', 'bandar', 'jackpot', 'scatter', 'gacor',
    'maxwin', 'rtp', 'pragmatic', 'pgsoft', 'livecasino', 'sportsbook',
    'roulette', 'blackjack', 'baccarat', 'sicbo', 'dragontiger',
    'slotgacor', 'judol', 'cuan',
    // Brand mirror / SEO clone (hostname mengandung nama brand)
    'sbobet', 'mansion88', 'm88', 'w88', 'fun88', 'we88', 'md88',
    'bk8', 'dafabet', 'cmd368', 'ibcbet', 'maxbet', '188bet',
    'longfu88', 'kawanslot', 'naga303', 'dewapoker', 'idnpoker',
    'joker123', 'mpo88', 'parimatch', 'paripesa', '1xbet', '1win',
    'mostbet', 'melbet', 'roobet', 'bcgame',
];

// Cache blocklist in memory, refresh periodically
let cachedDomains = [];
let lastRefresh = 0;
const CACHE_TTL = 60000; // 1 minute

async function refreshCache() {
    if (Date.now() - lastRefresh > CACHE_TTL) {
        try {
            cachedDomains = await getAllBlockedDomains();
            lastRefresh = Date.now();
        } catch (err) {
            console.warn('[JudiFilter] Failed to refresh cache:', err.message);
        }
    }
}

/**
 * Periksa apakah sebuah URL mengarah ke situs judi atau daftar blokir manual
 * @param {string} url - URL yang akan diperiksa
 * @returns {Promise<{ isBlocked: boolean, type: 'none' | 'judi' | 'manual' }>}
 */
async function checkUrlBlock(url) {
    if (!url) return { isBlocked: false, type: 'none' };

    await refreshCache();

    try {
        // Normalize URL
        let normalizedUrl = url;
        if (!normalizedUrl.startsWith('http://') && !normalizedUrl.startsWith('https://')) {
            normalizedUrl = 'https://' + normalizedUrl;
        }

        const parsedUrl = new URL(normalizedUrl);
        // Strip Bidi/zero-width marks + leading www.
        const host = parsedUrl.hostname
            .toLowerCase()
            .replace(/[\u200e\u200f\u202a-\u202e\u2066-\u2069\u200b]/g, '')
            .replace(/^www\./, '');

        // 1. Cek exact match dengan blocklist database (Blokir Manual)
        if (cachedDomains.includes(host)) return { isBlocked: true, type: 'manual' };

        // 2. Cek subdomain match (Blokir Manual)
        const parts = host.split('.');
        if (parts.length >= 2) {
            const baseDomain = parts.slice(-2).join('.');
            if (cachedDomains.includes(baseDomain)) return { isBlocked: true, type: 'manual' };

            if (parts.length >= 3) {
                const baseDomain3 = parts.slice(-3).join('.');
                if (cachedDomains.includes(baseDomain3)) return { isBlocked: true, type: 'manual' };
            }
        }

        // 3. Cek keyword di hostname (Deteksi Otomatis Judi)
        for (const keyword of JUDI_KEYWORDS) {
            if (host.includes(keyword)) return { isBlocked: true, type: 'judi' };
        }

        // 4. Cek keyword di path (Deteksi Otomatis Judi)
        const specificPathKeywords = ['slot', 'togel', 'casino', 'poker', 'sportsbook', 'livecasino'];
        const pathLower = parsedUrl.pathname.toLowerCase();
        for (const keyword of specificPathKeywords) {
            if (pathLower.includes(keyword)) return { isBlocked: true, type: 'judi' };
        }

        return { isBlocked: false, type: 'none' };
    } catch (e) {
        // Jika URL tidak valid, cek keyword di string mentah
        const urlLower = url.toLowerCase();
        const isJudi = JUDI_KEYWORDS.some(k => urlLower.includes(k));
        return { isBlocked: isJudi, type: isJudi ? 'judi' : 'none' };
    }
}

// Keep backward compatibility
async function isJudiSite(url) {
    const result = await checkUrlBlock(url);
    return result.isBlocked;
}

/**
 * Dapatkan daftar blocklist saat ini
 * @returns {Promise<{ domains: string[], keywords: string[], version: number }>}
 */
async function getBlocklist() {
    const domains = await getAllBlockedDomains();
    return {
        domains,
        keywords: JUDI_KEYWORDS,
        version: 1,
    };
}

/**
 * Tambah domain baru ke blocklist
 * @param {string} domain
 */
async function addToBlocklist(domain) {
    await insertBlockedDomain(domain.toLowerCase());
    lastRefresh = 0; // Force cache refresh
}

/**
 * Hapus domain dari blocklist
 * @param {string} domain
 */
async function removeFromBlocklist(domain) {
    await deleteBlockedDomain(domain.toLowerCase());
    lastRefresh = 0;
}

module.exports = {
    isJudiSite,
    checkUrlBlock,
    getBlocklist,
    addToBlocklist,
    removeFromBlocklist,
    JUDI_KEYWORDS,
};
